python/triton/compiler/compiler.py

```diff
@@ -9,32 +9,95 @@
 from ..runtime.cache import get_cache_manager, get_dump_manager, get_override_manager
 from ..runtime.driver import driver
 # TODO: this shouldn't be here
-from dataclasses import dataclass
 from .code_generator import ast_to_ttir
 from pathlib import Path
 import re
 import functools
 import os
 
 
-@dataclass
-class AttrsDescriptor:
-    divisible_by_16: set = None
-    equal_to_1: set = None
+# Common argument specializations
+def common_specializations(params, args):
+    arg_properties = {}
+
+    def is_divisible_by_16(x):
+        if hasattr(x, "data_ptr"):
+            return x.data_ptr() % 16 == 0
+        elif isinstance(x, int):
+            return x % 16 == 0
+        if x is None:
+            return True
+        return False
+
+    # Divisibility specialization
+    arg_properties["tt.divisibility"] = {
+        (param.num, 16)
+        for param, arg in zip(params, args)
+        if is_divisible_by_16(arg) and not param.do_not_specialize and not param.do_not_specialize_on_alignment
+    }
+
+    # Equal to 1 specialization
+    arg_properties["tt.equal_to_1"] = {
+        (param.num, 1)
+        for param, arg in zip(params, args)
+        if isinstance(arg, int) and not isinstance(arg, bool) and arg == 1 and not param.do_not_specialize
+    }
+    return arg_properties
 
-    def __post_init__(self):
-        if self.divisible_by_16 is None:
-            self.divisible_by_16 = set()
-        if self.equal_to_1 is None:
-            self.equal_to_1 = set()
 
-    def to_dict(self):
-        return {'divisible_by_16': list(self.divisible_by_16), 'equal_to_1': list(self.equal_to_1)}
+# Backend specific specializations
+def cuda_specializations(specializations, params, args):
+    return specializations
 
+
+def hip_specializations(specializations, params, args):
+    return specializations
```

Comment:
I don't understand the point of those? It seems to be doing the same thing. Is this where the specific specialization would happen?

Reply:
Yes, this is where I thought the specific specialization would happen. For now the are `nop`.